mitiq/shadows/quantum_processing.py

```diff
@@ -0,0 +1,181 @@
+    the circuit and return the measurement outcomes & the sampled
+    Pauli strings.
+
+    Args:
+         circuit: Cirq circuit.
+         n_total_measurements: number of snapshots.
+         sampling_function: Sampling function to use. If a string, then the
+            string is used to look up a default sampling function. If a
+            Callable, then the Callable is used as the sampling function.
+            Defaults to `"cirq"`.
+         sampling_function_config: Configuration
+            for the sampling function. Defaults to {}. If sampling_function is
+            a string, then this argument is ignored.
+
+    Returns:
+         outcomes: Tuple of two numpy arrays. The first array contains
+            measurement outcomes (-1, 1) while the second array contains the
+            index for the sampled Pauli's (0,1,2=X,Y,Z). Each row of the arrays
+            corresponds to a distinct snapshot or sample while each column
+            corresponds to a different qubit.
+    """
+
+    # Generate random Pauli unitaries
+    qubits = list(circuit.all_qubits())
+    num_qubits = len(qubits)
+    pauli_strings = generate_random_pauli_strings(
+        num_qubits, n_total_measurements
+    )
+    # Attach measurement gates to the circuit
+    rotated_circuits = tqdm(
+        get_rotated_circuits(
+            circuit,
+            pauli_strings,
+        ),
+        desc="Measurement",
+        leave=False,
+    )
+
+    if isinstance(sampling_function, str):
+        if sampling_function == "cirq":
+            # Run the circuits to collect the outcomes for cirq)
+            results = [
+                cirq_sample_bitstrings(
+                    rotated_circuit,
+                    noise_level=(0,),
+                    shots=1,
+                    sampler=cirq.Simulator(),
+                )
+                for rotated_circuit in rotated_circuits
+            ]
+        elif sampling_function == "qiskit":
+            # Run the circuits to collect the outcomes for cirq
+            results = [
+                qiskit_sample_bitstrings(
+                    to_qiskit(rotated_circuit),
+                    noise_model=None,
+                    backend=Aer.get_backend("aer_simulator"),
+                    shots=1,
+                    measure_all=False,
+                )
+                for rotated_circuit in rotated_circuits
+            ]
+        else:
+            raise ValueError(
+                f"Sampling function {sampling_function} not supported"
+            )
+    else:
+        results = [
+            sampling_function(rotated_circuit, **sampling_function_config)
+            for rotated_circuit in rotated_circuits
+        ]
+
+    # Transform the outcomes into a numpy array 0 -> 1, 1 -> -1.
+    shadow_outcomes = []
+    for result in results:
+        bitstring = list(result.get_counts().keys())[0]
+        outcome = [1 - int(i) * 2 for i in bitstring]
+        shadow_outcomes.append(outcome)
```

Comment:
Can you explain what you want to do here? Are you trying to get the bitstring which was measured the most during the sampling? Are you trying to always get the bitstring from the `00...0` state?

Reply:
yes, we need eigenvalue, 1, -1 instead of state vector |0>, |1>,

Reply:
Sorry Min, I need a bit more explanation. I'm asking you to explain the code.

As of python 3.7, dictionaries keep their "insertion ordering". So when getting the `0`th element of `list(result.get_counts().keys())`, are you expecting to get a particular bitstring? A random one? More explanation would be helpful.

Reply:
Hi Nate, thanks for the question. I understand that the line `bitstring = list(result.get_counts().keys())[0]` may cause some confusion. Let me explain it in detail below.
In clasical shadow, for each sampled random Pauli string (say "YXZ"), we only perform one-shot measurement. Hence, the `results.get_counts()` always has one key (which is the measurement outcome), and its value is always 1. Hence, to obtain the one-shot measurement outcome for this Pauli string, we just need to take the only key in `results.get_counts()` and convert it from string to integers, e.g., `"010" --> "[1, -1, 1"]`, where the bitstring "010" represents the state `|0>|1>|0>` and classical shadow needs their eigen-value representation (i.e., `"0" -> 1` and `"1" -> -1`).

Here is the printed results for `pauli_strings[0]`, `results[0]`, `results[0].get_counts())`:
```
YZX
MeasurementResult: {'nqubits': 3, 'qubit_indices': (0, 1, 2), 'shots': 1, 'counts': {'010': 1}}
{'010': 1}
```

Reply:
Thanks Nate for noticing this aspect and thanks Min for explaining that executors are expected to be called with a single shot. I had completely missed this fact.

I added a [suggestion](https://github.com/unitaryfund/mitiq/pull/1906#discussion_r1265164249) to explain this assumption in the function docstring.

Reply:
Thanks, Andrea, I have adopted this!